x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/task/gap_auto_fill_scheduler_task.ts

```diff
@@ -0,0 +1,387 @@
+              });
+              if (!capacityCheckInitial.canSchedule) {
+                await logEvent({
+                  status: GAP_AUTO_FILL_STATUS.SKIPPED,
+                  results: [],
+                  message: `Skipped execution: no capacity remaining to schedule gap fills (${capacityCheckInitial.currentCount}/${capacityCheckInitial.maxBackfills})`,
+                });
+                return { state: {} };
+              }
+
+              // Step 3: Fetch rule IDs with gaps
+              let remainingBackfills = capacityCheckInitial.remainingCapacity;
+              const startDate: Date | undefined = dateMath.parse(config.gapFillRange)?.toDate();
+              if (!startDate) {
+                throw new Error(`Invalid gapFillRange: ${config.gapFillRange}`);
+              }
+              // newest gap first
+              const sortOrder = 'desc';
+              const { ruleIds } = await rulesClient.getRuleIdsWithGaps({
+                start: startDate.toISOString(),
+                end: now.toISOString(),
+                sortOrder,
+              });
+
+              if (!ruleIds.length) {
+                await logEvent({
+                  status: GAP_AUTO_FILL_STATUS.SKIPPED,
+                  results: [],
+                  message: 'Skipped execution: no rules with gaps',
+                });
+                return { state: {} };
+              }
+
+              const aggregatedByRule = new Map<
+                string,
+                {
+                  ruleId: string;
+                  processedGaps: number;
+                  status: GapFillSchedulePerRuleStatus;
+                  error?: string;
+                }
+              >();
+
+              // Step 4: Process rules in batches
+              for (
+                let startIdx = 0;
+                startIdx < ruleIds.length;
+                startIdx += DEFAULT_RULES_BATCH_SIZE
+              ) {
+                // Stop early if we've reached capacity
+                if (remainingBackfills <= 0) {
+                  const consolidated = resultsFromMap(aggregatedByRule);
+                  await logEvent({
+                    status: GAP_AUTO_FILL_STATUS.SUCCESS,
+                    results: consolidated,
+                    message: `Stopped early: no backfill capacity remaining | ${formatConsolidatedSummary(
+                      consolidated
+                    )}`,
+                  });
+                  return { state: {} };
+                }
+                if (
+                  await handleCancellation({
+                    wasCancelled,
+                    abortController,
+                    aggregatedByRule,
+                    logEvent,
+                  })
+                ) {
+                  return { state: {} };
+                }
+
+                const currentRuleIds = ruleIds.slice(startIdx, startIdx + DEFAULT_RULES_BATCH_SIZE);
+
+                const { data: rules } = await rulesClient.find({
```

Comment:
A comment here would help to describe why we need to find the rules by ID just to extract the ID from each rule. Is it to reuse the authorization check in `findRules`? Are the IDs we're querying for and fetching not actually the same field? Filtering out rules that don't exist anymore?